server/statshouse/statshouse-client.cpp

```diff
@@ -8,20 +8,71 @@
 #include <unistd.h>
 
 #include "common/resolver.h"
+#include "common/server/stats.h"
+#include "common/stats/provider.h"
 #include "common/tl/constants/statshouse.h"
 #include "common/tl/methods/string.h"
 #include "net/net-connections.h"
 #include "runtime/critical_section.h"
 #include "server/server-log.h"
 
+class statshouse_stats_t : public stats_t {
+public:
+  statshouse_stats_t(const std::vector<std::pair<std::string, std::string>> &tags)
+    : tags(tags) {}
+
+  void add_general_stat(const char *, const char *, va_list) noexcept final {
+    // ignore it
+  }
+
+  void add_stat(char type [[maybe_unused]], const char *key, const char *value_format [[maybe_unused]], double value) noexcept final {
+    auto metric = make_statshouse_value_metric(normalize_key(key, "%s_%s", stats_prefix), value, tags);
+    auto len = vk::tl::store_to_buffer(sb.buff + sb.pos, sb.size, metric);
+    sb.pos += len;
+    ++counter;
+  }
+
+  void add_stat(char type, const char *key, const char *value_format, long long value) noexcept final {
+    add_stat(type, key, value_format, static_cast<double>(value));
+  }
+
+  virtual bool needAggrStats() noexcept final {
+    return false;
+  }
+
+  int get_counter() const {
+    return counter;
+  }
+
+private:
+  int counter{0};
+  const std::vector<std::pair<std::string, std::string>> &tags;
+};
+
+namespace {
+std::pair<char *, int> prepare_statshouse_stats(statshouse_stats_t &&stats, const char *statsd_prefix, unsigned int tag_mask) {
```

Comment:
You've renamed the field `statsd_prefix` to `stats_prefix`, but here is still `statsd_prefix`. In a function preparing statshouse).
Btw, I'd suggest to remove this parameter at all and just to hardcode "kphp" below